### ISSUE TITLE
Game freeze

### ISSUE THREAD

**Issue by wrkncacnter7** [Aleph-One-Marathon/alephone#182]
While playing a network game, the game froze/crashed for basically everyone, however, I was able to alt+q to get to the PCR and save the film. Playing back the film reproduces the freeze, though I have to force kill alephone once it happens.

Hopefully this film can be used to debug the issue. I believe the map is from http://simplici7y.com/items/caustic-dystopia.

[freeze_bug.zip](https://github.com/Aleph-One-Marathon/alephone/files/4536465/freeze_bug.zip)

**Comment by TrajansRow** [comment]
The renderer is getting stuck in a spin loop and cannot exit sort_render_tree(). I've never seen anything like this before!

**Comment by LidMop** [comment]
Yup, and it's because the render tree got corrupted (became cyclic) by build_render_tree(). It can do that when the view origin is sitting _exactly_ on a vertex.

Bungie evidently knew about this and added [this hack](https://github.com/Aleph-One-Marathon/alephone/blob/a61e96d8f54f4128519e7a29a3dad1c09cdd4c34/Source_Files/RenderMain/render.cpp#L640-L665) that tries to move off of a vertex a tiny bit just before the render tree is built, but it doesn't work in some cases. In particular, it doesn't work for any vertex that is equidistant from and colinear with its two neighboring vertices: `inset_vector` will be {0,0} and the origin will remain sitting on the vertex. The film triggers this case (incidentally, the offending polygon in the film is also a degenerate (zero-area) polygon, but that's a red herring).

For other geometries, the above hack can potentially shift the view origin out of the polygon, but it doesn't update `view->origin_polygon_index`. I'm working to confirm whether this is another case that doesn't work.

We could try to make Bungie's hack more robust (and complicated), but I would prefer to get rid of it entirely and just fix build_render_tree() directly, once I understand how it works.

**Comment by treellama** [comment]
There's a spot in OK Honeybunny where this can happen as well, I believe. There was an ancient bug in the SourceForge tracker about it.

**Comment by LidMop** [comment]
Ah yeah, [376](https://sourceforge.net/p/marathon/bugs/376/).

**Comment by treellama** [comment]
I had been meaning to refactor the vis tree building anyway. It's a bottleneck on complex maps, and a good candidate for parallelization. Whether I get time to do it will be another story.

**Comment by LidMop** [comment]
I reverted d707b20 due to a rendering regression and shored up Bungie's hack instead, as a stop-gap fix until I work out a fix for that regression.
